In [39]:
import numpy as np
import pandas as pd

from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score, precision_score, recall_score

In [28]:
file_path = "../../data/SVM/sonar.all-data.csv"

data = pd.read_csv(file_path, header=None)

data.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  
3  0.0040  0.0117   R  
4  0.0107  0.0094   R  

[5 rows x 61 columns]

In [ ]:
X = data.iloc[:, :60]
y = data.iloc[:, -1]

In [30]:
y = y.map({"R": -1, "M": 1})

In [ ]:
clf = SVC(kernel='linear', C = 1e6)

clf.fit(X, y)

y_pred = clf.predict(X)

print("Accuracy:", accuracy_score(y, y_pred))
print("Precision:", precision_score(y, y_pred))
print("Recall:", recall_score(y, y_pred))

Accuracy: 1.0
Precision: 1.0
Recall: 1.0


In [45]:
clf = SVC(kernel='linear', C = 3.5)

clf.fit(X, y)

y_pred = clf.predict(X)

print("Accuracy:", accuracy_score(y, y_pred))
print("Precision:", precision_score(y, y_pred))
print("Recall:", recall_score(y, y_pred))

Accuracy: 0.8509615384615384
Precision: 0.8333333333333334
Recall: 0.9009009009009009


### Nhận xét:

* Trường hợp C >= 10^5:
    * Accuracy = 1 
    * Precision = 1
    * Recall = 1

    => Mô hình có độ chính xác tuyệt đối, nhưng khả năng overfitting rất cao.

* Trường hợp C trong [0.5, 5], chọn C = 3.5:
    * Accuracy = 0.8509615384615384
    * Precision = 0.8333333333333334
    * Recall = 0.9009009009009009
    
    => Mô hình đạt hiệu suất tốt và cân bằng, phù hợp cho tổng quát hóa.

In [43]:
parameters = {'kernel': ['linear'], 'C':[1e5, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5]}
svc = SVC()
clf = GridSearchCV(svc, parameters)
clf.fit(X, y)

GridSearchCV(estimator=SVC(),
             param_grid={'C': [100000.0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5,
                               5],
                         'kernel': ['linear']})